sklearn/neighbors/quad_tree.pyx

```diff
@@ -0,0 +1,588 @@
+            printf("[QuadTree] Checking point (%f, %f, %f) in cell %li "
+                    "([%f/%f, %f/%f, %f/%f], size %li)\n",
+                    point[0], point[1], point[2], cell.cell_id,
+                    cell.min_bounds[0], cell.max_bounds[0], cell.min_bounds[1],
+                    cell.max_bounds[1], cell.min_bounds[2], cell.max_bounds[2],
+                    cell.cumulative_size)
+
+        for i in range(self.n_dimensions):
+            if (cell.min_bounds[i] > point[i] or
+                    cell.max_bounds[i] <= point[i]):
+                with gil:
+                    msg = "[QuadTree] InsertionError: point out of cell "
+                    msg += "boundary.\nAxis %li: cell [%f, %f]; point %f\n"
+
+                    msg %= i, cell.min_bounds[i],  cell.max_bounds[i], point[i]
+                    raise ValueError(msg)
+
+    def _check_coherence(self):
+        """Check the coherence of the cells of the tree.
+
+        Check that the info stored in each cell is compatible with the info
+        stored in descendent and sibling cells. Raise a ValueError if this
+        fails.
+        """
+        for cell in self.cells[:self.cell_count]:
+            # Check that the barycenter of inserted point is within the cell
+            # boundaries
+            self.check_point_in_cell(cell.barycenter, &cell)
+
+            if not cell.is_leaf:
+                # Compute the number of point in children and compare with
+                # its cummulative_size.
+                n_points = 0
+                for idx in range(self.n_cells_per_cell):
+                    child_id = cell.children[idx]
+                    if child_id != -1:
+                        child = self.cells[child_id]
+                        n_points += child.cumulative_size
+                        assert child.cell_id == child_id, (
+                            "Cell id not correctly initiliazed.")
+                if n_points != cell.cumulative_size:
+                    raise ValueError(
+                        "Cell {} is incoherent. Size={} but found {} points "
+                        "in children. ({})"
+                        .format(cell.cell_id, cell.cumulative_size,
+                                n_points, cell.children))
+
+        # Make sure that the number of point in the tree correspond to the
+        # cummulative size in root cell.
+        if self.n_points != self.cells[0].cumulative_size:
+            raise ValueError(
+                "QuadTree is incoherent. Size={} but found {} points "
+                "in children."
+                .format(self.n_points, self.cells[0].cumulative_size))
+
+    cdef long summarize(self, DTYPE_t[3] point, DTYPE_t* results,
+                        SIZE_t cell_id=0, long idx=0, float squared_theta=.5
+                        ) nogil:
+        """Summarize the tree compared to a query point.
+
+        Input arguments
+        ---------------
+        point: array (n_dimensions)
```

Comment:
Should really have spaces before colons for valid RST/numpydoc (even if the headings aren't to standard)